File: problem2.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

//  Two HashMaps, sMap and tMap, are used to map characters from
//  string s to characters in string t and vice versa.

import java.util.HashMap;

class problem2 {
    public boolean isIsomorphic(String s, String t) {
        if(s == null && t == null) return true;
        if(s== null || t ==null) return false;
        if(s.length() != t.length()) return false;
        HashMap<Character, Character> sMap = new HashMap<>();
        HashMap<Character, Character> tMap = new HashMap<>();

        for(int i = 0; i < s.length(); i++) {
            char sChar = s.charAt(i);
            char tChar = t.charAt(i);
            if(!sMap.containsKey(sChar)) {
                sMap.put(sChar, tChar);
            } else {
                if(sMap.get(sChar) != tChar) {
                    return false;
                }
            }

            if(!tMap.containsKey(tChar)){
                tMap.put(tChar, sChar);
            } else {
                if(tMap.get(tChar) != sChar){
                    return false;
                }
            }
        }
        return true;
    }
}

File: problem-2.go
/* time complexity : O(n)
space complexity : O(n) */

func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	sMap := make(map[byte]byte)
	tMap := make(map[byte]byte)
	for i := 0; i < len(s); i++ {
		sVal := s[i]
		tVal := t[i]
		if val, ok := sMap[sVal]; ok {
			if val != tVal {
				return false
			}
		} else if _, ok := tMap[tVal]; ok {
			return false
		}
		sMap[sVal] = tVal
		tMap[tVal] = sVal
	}
	return true
}